src/main/java/net/minecraftforge/event/world/JigsawPatternInitEvent.java

```diff
@@ -0,0 +1,285 @@
+                if (this.newCategories.size() != 0)
+                {
+                    AtomicInteger i = new AtomicInteger(0);
+                    stringBuilder.append("\tNew Categories: ").append('\n');
+                    this.newCategories.keySet().forEach(resourceLocation ->
+                    {
+                        stringBuilder.append("\t\tEntry: ").append(i.getAndIncrement()).append(", ").append(resourceLocation).append('\n');
+                    });
+                }
+                if (this.remove.size() != 0)
+                {
+                    stringBuilder.append("\tRemoved JigsawPieces from Category").append('\n');
+                    this.remove.entrySet().forEach((resourceLocationListEntry ->
+                    {
+                        AtomicInteger i = new AtomicInteger(0);
+                        stringBuilder.append("\t\tFrom Category: ").append(resourceLocationListEntry.getKey()).append('\n');
+                        resourceLocationListEntry.getValue().forEach(resourceLocation ->
+                        {
+                            stringBuilder.append("\t\t\tEntry: ").append(i.getAndIncrement()).append(", ").append(resourceLocation).append('\n');
+                        });
+                    }));
+                }
+                if (this.newBuildings.size() != 0)
+                {
+                    stringBuilder.append("\tAdded JigsawPieces to Category").append('\n');
+                    this.newBuildings.entrySet().forEach((resourceLocationListEntry ->
+                    {
+                        AtomicInteger i = new AtomicInteger(0);
+                        stringBuilder.append("\t\tTo Category: ").append(resourceLocationListEntry.getKey()).append('\n');
+                        resourceLocationListEntry.getValue().forEach(pair ->
+                        {
+                            stringBuilder.append("\t\t\tEntry: ").append(i.getAndIncrement()).append(", ").append(pair.getFirst().getRegistryName()).append('\n');
+                        });
+                    }));
+                }
+            }));
+        }
+    }
+
+    /**
+     * This event should be fired whenever a Jigsaw bases structure initializes its pools.
+     * Use this event to register custom JigsawPattern before the corresponding structure is constructed.
+     */
+    public static abstract class StructureJigsawPoolInitEvent extends Event
+    {
+
+        public void register(JigsawPattern jigsawPattern)
+        {
+            JigsawManager.field_214891_a.register(jigsawPattern);
+        }
+
+        public abstract ResourceLocation getStructureRegistryName();
+
+        public static class Village extends StructureJigsawPoolInitEvent{
+            private static boolean fired = false;
+
+            public static void fire()
+            {
+                if(!fired)
+                {
+                    MinecraftForge.EVENT_BUS.post(new Village());
+                    fired = true;
```

Comment:
Why does this use a static varibale to fire once? Also, this belongs in ForgeHooks or ForgeEventFactory